library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,261 @@
+//! as a "sandbox" that cannot be broken into or out of. *Provenance* is the permission
+//! to access an allocation's sandbox and has both a *spatial* and *temporal* component:
+//!
+//! * Spatial: A range of bytes in the allocation that the pointer is allowed to access.
+//! * Temporal: Some kind of globally unique identifier tied to the allocation itself.
+//!
+//! Spatial provenance makes sure you don't go beyond your sandbox, while temporal provenance
+//! makes sure that you can't "get lucky" after your permission to access some memory
+//! has been revoked (either through deallocations or borrows expiring).
+//!
+//! Provenance is implicitly shared with all pointers transitively derived from
+//! The Original Pointer through operations like [`offset`], borrowing, and pointer casts.
+//! Some operations may *shrink* the derived provenance, limiting how much memory it can
+//! access or how long it's valid for (i.e. borrowing a subfield and subslicing).
+//!
+//! Shrinking provenance cannot be undone: even if you "know" there is a larger allocation, you
+//! can't derive a pointer with a larger provenance. Similarly, you cannot "recombine"
+//! two contiguous provenances back into one (i.e. with a `fn merge(&[T], &[T]) -> &[T]`).
+//!
+//! A reference to a value always has provenance over exactly the memory that field occupies.
+//! A reference to a slice always has provenance over exactly the range that slice describes.
+//!
+//! If an allocation is deallocated, all pointers with provenance to that allocation become
+//! invalidated, and effectively lose their provenance.
+//!
+//! The strict provenance experiment is mostly only interested in exploring stricter *spatial*
+//! provenance. In this sense it can be thought of as a subset of the more ambitious and
+//! formal [Stacked Borrows][] research project, which is what tools like [miri][] are based on.
+//! In particular, Stacked Borrows is necessary to properly describe what borrows are allowed
+//! to do and when they become invalidated. This necessarily involves much more complex
+//! *temporal* reasoning than simply identifying allocations.
+//!
+//!
+//! ## Pointer Vs Addresses
+//!
+//! **This section is *non-normative* and is part of the [Strict Provenance][] experiment.**
+//!
+//! One of the largest historical issues with trying to define provenance is that programmers
+//! freely convert between pointers and integers. Once you allow for this, it generally becomes
+//! impossible to accurately track and preserve provenance information, and you need to appeal
+//! to very complex and unreliable heuristics. But of course, converting between pointers and
+//! integers is very useful, so what can we do?
+//!
+//! Strict Provenance attempts to square this circle by decoupling Rust's traditional conflation
+//! of pointers and `usize` (and `isize`), and defining a pointer to semantically contain the
+//! following information:
+//!
+//! * The **address-space** it is part of.
```

Comment:
Address spaces kind of come out of the blue here, this is confusing.

Also, isn't handling address spaces simple -- you just encode in the type of the pointer which address space it refers to?

Reply:
I want to be intentionally vague here, if that's ok. (I think you can have address-spaces which are burned into the pointer's repr, but this is... beyond my knowledge).

Reply:
Well right now it's confusing since all the text before talks about how provenance is what makes pointers complicated, and then suddenly this "address space" stuff also shows up without any context.

Reply:
You are definitely right. Hrm.

Reply:
Basically this part is in a bit of a zombie-state because I am flip-flopping on how much I want to *explicitly* talk about segmentation and how much I want to vaguely allude to it. Because priority-wise segmentation is like... quartenary for me?

Reply:
Would it be an option to simply not mention address spaces and implicitly treat them as part of 'provenance'?

Reply:
The latest push addresses this, I think: Just beforehand I bring up the problem that WASM is actually a harvard architecture but we still just say it's totally fine to cast between data and fn ptrs, which at very least is in the same wheelhouse as "pointer casts that are dubious". I think that sets up enough motivation for suddenly introducing the notion of address-spaces and caring about them.

Reply:
Okay, I like that. However I think the connection should be made explicit, either by mention "address spaces" in the WASM section or by mentioning the WASM function pointer situation here.

Reply:
changed to

```
//! * The **address-space** it is part of (i.e. "data" vs "code" in WASM).
```